rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,351 @@
+The Google-internal test for TF should catch the error if the rollback sequence
+is not correct.
+
+```python
+# Existing scenario.
+# PR 11111 (2 files updated)
+# +++ tensorflow/python/ops/array_ops.py
+def some_new_function(inputs):
+   ...
+
+# +++ tensorflow/python/keras/layers/core.py
+
+class new_layer(Layer):
+
+  def call(inputs):
+     array_ops.some_new_function(inputs)
+     ...
+```
+
+```python
+# New scenario.
+# PR 22222 (1 file updated)
+# +++ tensorflow/python/ops/array_ops.py
+@tf.export('some_new_function')
+def some_new_function(inputs):
+   ...
+
+==================================
+# PR 33333 (1 file updated)
+# +++ tensorflow/python/keras/layers/core.py
+
+class new_layer(Layer):
+
+  def call(inputs):
+     tf.some_new_function(inputs)
+     ...
+```
+
+2. Changing the behavior of an existing TF API.
+
+Note that the PR 22222 needs to be submitted with both the new and old
+function since Google internal CI is still testing from HEAD.
+The previous function can be 
+deleted after PR 33333 is submitted. Also note that this issue is caused by 
+Keras not using exclusively public TF API, but relying on TF implementation details.
+Moving towards only using public APIs should reduce the likelihood of this kind of issue.
+
+```python
+# Existing scenario.
+# PR 11111 (2 files updated)
+# tensorflow/python/ops/array_ops.py
+<<<
+def existing_function(inputs):
+    ...
+>>>
+def new_function(inputs, knob1=False, knob2=1):
+    ...
+# tensorflow/python/keras/layers/core.py
+
+class existing_layer(Layer):
+
+  def call(inputs):
+<<<
+    array_ops.existing_function(inputs)
+>>>
+    array_ops.new_function(
+        inputs, 
+        knob1=True,
+        knob2=3)
+```
+
+```python
+# New scenario.
+# PR 22222 (1 file updated)
+# tensorflow/python/ops/array_ops.py
+<<<
+def existing_function(inputs):
+   ...
+>>>
+def existing_function(inputs):
+  return new_function(
+    inputs, 
+    knob1=False,
+    knob2=1)
+
+def new_function(inputs, knob1, knob2=1):
+    ...
+
+==================================
+# PR 33333 (1 file updated)
+# tensorflow/python/keras/layers/core.py
+class existing_layer(Layer):
+
+  def call(inputs):
+<<<
+    array_ops.existing_function(inputs)
+     ...
+>>>
+    array_ops.new_function(
+        inputs, 
+        knob1=True,
+        knob2=3)
+```
+
+
+### Performance Implications
+
+There may be some performance implications as we move towards only using
+public TF APIs. We need to maintain a benchmark to ensure that there
+is no performance regression.
+
+### Dependencies
+
+The TensorFlow pip package will auto-install the Keras package, which shouldn't make 
+any difference on the end-user side. Under the hood, Keras will be a different 
+package imported by `tf_core`, like what we do for TF estimator.
+
+### Developer experience Impact
+
+* The local build and test times should be greatly reduced, since compiling TF
+is no longer needed, and Keras is a pure-Python project.
```

Comment:
The build times will be greatly reduced, but Keras is unlikely to remain pure-python forever.

We may also experience a future where Keras contains custom ops that aren't in core TF (similarly to Addons).
Right now that's very painful to sync internally & externally because the custom ops are handled in a different way, so we may end up having to work on improving that.

Reply:
Done. updated wording for pure python project.